[Concept Entry] Rust Strings (#4251)
Codecademy/docs: content/rust/concepts/strings/strings.md

---
Title: "Strings"

> Description: "This concise guide covers Rust's distinctive treatment of strings, exploring types like `String` and `&str`, syntax details, and string manipulation using operators (`+`, `+=`, `format!`)."

Reviewer: make this short and precise desc of strings. Don't have to be this long.

Subjects:
  - "Computer Science"

>   - "Code Foundations"

Author: the double used to throw an error so I changed it, but I'll try it again.

CatalogContent:
  - "concepts/slices/sliced.md"
  - "paths/computer-science"
---

> 
> 
> ## Strings

Reviewer: ```suggestion

```

> **Strings** in Rust diverge a little bit from conventional programming languages. The approach to string manipulation in Rust is influenced by its systems-focused design. This concept entry will provide insights into creating and manipulating strings in Rust, offering practical examples to illustrate key aspects. 

Reviewer: ```suggestion
**`Strings`** in Rust diverge a little bit from conventional programming languages. The approach to string manipulation in Rust is influenced by its systems-focused design. This concept entry will provide insights into creating and manipulating strings in Rust, offering practical examples to illustrate key aspects. 
```

> Tackling data structures of variable size, such as strings, can get a bit tricky, and Rust has its own spin on it. In Rust, a string is essentially a sequence of Unicode characters encoded in UTF-8. As an example - the string "Codecademy Rust concept," where each character is a valid Unicode entity – "C," "o," "d," "e," "c," and so on.**

Reviewer: ```suggestion
Tackling data structures of variable size, such as strings, can get a bit tricky, and Rust has its spin on it. In Rust, a string is essentially a sequence of Unicode characters encoded in UTF-8. As an example - the string "Codecademy Rust concept," where each character is a valid Unicode entity – "C," "o," "d," "e," "c," and so on.
```

> ### String types

Reviewer: ```suggestion
## String types
```

  
In Rust, there are two main types related to strings, each serving a specific purpose.
1. **`String`**
   - **Description:** A heap-allocated, growable, and mutable string type.
   - **Usage:** Used when dynamic string manipulation and ownership transfer are required.
   - **Example:**
     ```rust
     let mut dynamic_string = String::from("Hello, Rust!");
     ```

2. **`&str` (String Slice)**
   - **Description:** A reference to a sequence of UTF-8 bytes, often used as a view into a `String` or a string literal.
   - **Usage:** Commonly used for string references without ownership or when dealing with parts of a string.
   - **Example:**
     ```rust
     let string_literal: &str = "Hello, Rust!";
     ```

> > [!NOTE]
> > These string types cover various scenarios, from dynamic and mutable strings `String` to static and immutable string slices `&str`. The type of string should be chosen depending on specific requirements and characteristics of the data being manipulated.

Reviewer: ```suggestion
> **Note:** These string types cover various scenarios, from dynamic and mutable strings `String` to static and immutable string slices `&str`. The type of string should be chosen depending on the specific requirements and the characteristics of the data being manipulated.
```

>   ***

Reviewer: ```suggestion
```

>     
> ## Syntax with Examples
> 
> There are three categories to understand the syntax of Rust strings: *creating, manipulating, and referencing* strings.

Reviewer: ```suggestion
```

Let's remove this and create the others as headings.

Author: That was the idea - having an intro sentence with the three topics as headlines underneath. I'll adjust it :)

> ### Creating Strings

Reviewer: ```suggestion
## Creating Strings
```

1. **Using `String::new()`**:

   ```rust
   let mut empty_string = String::new();
   ```

   This creates a new, mutable, empty `String` that can be later modified and or assigned to a variable.

2. **Using String Literals with `String::from`**:

   ```rust
   let greeting = String::from("Hello, World!");
   ```

   `String::from` allocates memory on the heap and initializes a new string with the specified string literal.

> ### String Manipulation

Reviewer: ```suggestion
## String Manipulation
```

   - **Example:**
     ```rust
     let hello = String::from("Hello, ");
     let world = String::from("World!");
     let hello_world = hello + &world;  // here akes ownership of 'hello'
     ```

   > **`+=` Operator:**

     - **Usage:** in-place concatenation operation
     - **Ownership:** appends to the existing string
     - **Borrowing:** It works directly with the mutable reference of the left operand.
     - **Example:**
       ```rust
       let mut hello = String::from("Hello, ");
       let world = String::from("World!");
       hello += &world;  // Appends 'World!' to the existing 'Hello, '
       ```

   > **`format!` Macro:**

     - **Usage:** creates a new string by formatting text, allowing string interpolation.
     - **Ownership:** It does not involve ownership transfer and is a convenient way to create strings without borrowing or ownership concerns.
     - **Example:**
       ```rust
       let hello = String::from("Hello, ");
       let world = String::from("World!");
       let hello_world = format!("{}{}", hello, world);  // Creates a new string without ownership issues
       ```

>   > [!IMPORTANT]
>   > When working with strings in Rust, it's crucial to be mindful of ownership and borrowing semantics, especially when using operators like `+`. The operator `+` is used to create a new string while  and `+=` is used to modify an existing string in place. To concatenate strings with interpolation the `format!` macro provides a flexible and ownership-friendly way.

Reviewer: ```suggestion
> **Note:** When working with strings in Rust, it's crucial to be mindful of ownership and borrowing semantics, especially when using operators like `+`. The operator `+` is used to create a new string while `+=` is used to modify an existing string in place. To concatenate strings with interpolation the `format!` macro provides a flexible and ownership-friendly way.
```

> ---

Reviewer 2: The metadata is not formatted correctly. It should be formatted like this, the line spaces, characters should be strictly like the below example - 

---
Title: 'The Title'
Description: 'A brief description'
Subjects:
  - 'A subject name'
  - 'A second subject name'
  - 'An nth subject name'
Tags:
  - 'A tag'
  - 'A second tag'
  - 'An nth tag'
CatalogContent:
  - 'learn-example-course'
  - 'path/example-path'
---

Title: 'Strings'

> Description: 'This entry covers creating, manipulating and referencing strings (encodings of UTF-8 sequences) in Rust.'

Reviewer 2: Describe what are strings in Rust rather than using this sentence. Write a one-line description.

Subjects:

> 

Reviewer 2: ```suggestion
```

- 'Computer Science'
- 'Code Foundations'
  CatalogContent:
- 'learn-rust'
- 'paths/computer-science'

> 

Reviewer 2: ```suggestion
```

---

**`Strings`** in Rust diverge a little bit from conventional programming languages. The approach to string manipulation
in Rust is influenced by its systems-focused design. This concept entry will provide insights into creating and
manipulating strings in Rust, offering practical examples to illustrate key aspects.

Tackling data structures of variable size, such as strings, can get a bit tricky, and Rust has its spin on it. In Rust,
a string is essentially a sequence of Unicode characters encoded in UTF-8. As an example - the string "Codecademy Rust

> concept," where each character is a valid Unicode entity – "C," "o," "d," "e," "c," and so on.

Reviewer 2: Again describe what are strings in rust and other aspects of it. Dont write lines like these - 
`This concept entry will provide insights into creating and manipulating strings in Rust, offering practical examples to illustrate key aspects.`. Modify the introduction.

## String types

In Rust, there are two main types related to strings, each serving a specific purpose.

1. **`String`**

   - **Description:** A heap-allocated, growable, and mutable string type.
   - **Usage:** Used when dynamic string manipulation and ownership transfer are required.
   - **Example:**

     ```rust
     let mut dynamic_string = String::from("Hello, Rust!");
     ```

2. **`&str` (String Slice)**

   - **Description:** A reference to a sequence of UTF-8 bytes, often used as a view into a `String` or a string
     literal.
   - **Usage:** Commonly used for string references without ownership or when dealing with parts of a string.
   - **Example:**

     ```rust
     let string_literal: &str = "Hello, Rust!";
     ```

> > **Note:** These string types cover various scenarios, from dynamic and mutable strings `String` to static and
> > immutable string slices `&str`. The type of string should be chosen depending on the specific requirements and the
> > characteristics of the data being manipulated.

Reviewer 2: ```suggestion
> **Note:** These string types cover various scenarios, from dynamic and mutable strings `String` to static and immutable string slices `&str`. The type of string should be chosen depending on the specific requirements and the characteristics of the data being manipulated.
```

## Creating Strings

> 1. **Using `String::new()`**:
> 
>    ```rust
>    let mut empty_string = String::new();
>    ```
> 
>    This creates a new, mutable, empty `String` that can be later modified and or assigned to a variable.
> 
> 2. **Using String Literals with `String::from`**:
> 
>    ```rust
>    let greeting = String::from("Hello, World!");
>    ```
> 
>    `String::from` allocates memory on the heap and initializes a new string with the specified string literal.

Reviewer 2: ```suggestion
1. **Using `String::new()`**:

   ```rust
   let mut empty_string = String::new();
   ```

   This creates a new, mutable, empty `String` that can be later modified and or assigned to a variable.

2. **Using String Literals with `String::from()`**:

   ```rust
   let greeting = String::from("Hello, World!");
   ```

   `String::from()` allocates memory on the heap and initializes a new string with the specified string literal.
```

Author: @avdhoottt Please explain what you want me to do here. I can't seem to find a difference/the comment seems to be cut in half.

Author: > @avdhoottt Please explain what you want me to do here. I can't seem to find a difference/the comment seems to be cut in half.

@avdhoottt still waiting for an answer here. :)

## String Manipulation

> 1. **Concatenation**

Reviewer 2: You can use H3 for these headings. For example, Concatenation, Slicing & Appending with `push_str` and `push`.

Author: @avdhoottt Thank you for your comments! 
I did have those as h3 before, but changed them as suggested in a change request before.. would this be the final change request regarding that topic or should this be discussed? I just want to prevent more changes and commits to be done back and forth. :)

Author: Since I haven't heard from you except for your comment from the 6th of February, I will implement this again.

>    In Rust, there are multiple ways to concatenate strings depending on the given strings. Here is a list of
>    concatenation operators including examples:

Reviewer 2: ```suggestion
   In Rust, there are multiple ways to concatenate strings. Here is a list of concatenation operators including examples:
```

>    > **`+` Operator:**

Reviewer 2: ```suggestion
   **`+` Operator:**
```
Don't use `>` as this is used to convey a note. So please remove that from each heading below.

>    - **Usage:** concatenating two strings
>    - **Ownership:** takes ownership of the left operand
>    - **Borrowing:** To concatenate with a borrowed string (sliced out of another string using `&str`), it is necessary
>      to explicitly borrow it using the `&` operator.
>    - **Example:**
> 
>      ```rust
>      let hello = String::from("Hello, ");
>      let world = String::from("World!");
>      let hello_world = hello + &world;  // here takes ownership of 'hello'
>      ```

Reviewer 2: ```suggestion
   - **Usage:** Concatenating two strings.
   - **Ownership:** Consumes the left operand.
   - **Borrowing:** To concatenate with a borrowed string (sliced out of another string using `&str`), it is necessary to explicitly borrow it using the `&` operator.
   - **Example:**

     ```rust
     let hello = String::from("Hello, ");
     let world = String::from("World!");
     let hello_world = hello + &world;  // Ownership of 'hello' is moved, 'world' is borrowed
     ```
```

   > **`+=` Operator:**

>    - **Usage:** in-place concatenation operation
>    - **Ownership:** appends to the existing string
>    - **Borrowing:** It works directly with the mutable reference of the left operand.
>    - **Example:**

Reviewer 2: ```suggestion
   - **Usage:** In-place concatenation operation.
   - **Ownership:** Consumes the existing string on the left operand and appends to it.
   - **Borrowing:** It works directly with the mutable reference of the left operand, but the ownership of the left operand is transferred to the result.
   - **Example:**
```

     ```rust
     let mut hello = String::from("Hello, ");
     let world = String::from("World!");
     hello += &world;  // Appends 'World!' to the existing 'Hello, '
     ```

>    > **`format!` Macro:**
> 
>    - **Usage:** creates a new string by formatting text, allowing string interpolation.
>    - **Ownership:** It does not involve ownership transfer and is a convenient way to create strings without borrowing
>      or ownership concerns.
>    - **Example:**
> 
>      ```rust
>      let hello = String::from("Hello, ");
>      let world = String::from("World!");
>      let hello_world = format!("{}{}", hello, world);  // Creates a new string without ownership issues
>      ```

Reviewer 2: ```suggestion
   > **`format!` Macro:**

   - **Usage:** Creates a new string by formatting text, allowing string interpolation.
   - **Ownership:** It does not involve ownership transfer and is a convenient way to create strings without borrowing
     or ownership concerns.
   - **Example:**

     ```rust
     let hello = String::from("Hello, ");
     let world = String::from("World!");
     let hello_world = format!("{}{}", hello, world);  // Creates a new string without ownership issues
     ```
```

>    > **Note:** When working with strings in Rust, it's crucial to be mindful of ownership and borrowing semantics,
>    > especially when using operators like `+`. The operator `+` is
>    > used to create a new string while `+=` is used to modify an existing string in place. To concatenate strings with
>    > interpolation the `format!` macro provides a flexible
>    > and ownership-friendly way.

Reviewer 2: ```suggestion
   > **Note:** When working with strings in Rust, it's crucial to be mindful of ownership and borrowing semantics, especially when using operators like `+`. The operator `+` is used to create a new string while `+=` is used to modify an existing string in place. To concatenate strings with interpolation the `format!` macro provides a flexible and ownership-friendly way.
```

2. **Slicing & Appending with `push_str` and `push`**:

   ```rust
   let mut message = String::from("Rust");
   message.push_str(" Programming");
   message.push('!');
   let part_of_message = &message[0..5];
   ```

>    In the example above, `push_str` is utilized to append a string slice to the existing string `message`, and `push` is
>    used to add the character `!` to the end of the concatenated string.

Reviewer 2: ```suggestion
   In the example above, `push_str` is used to append a string slice to the existing string `message`, and `push` is used to add the character `!` to the end of the concatenated string.
```

## Referencing Strings

To reference parts of a string without ownership `&str` can be used as follows:

```rust
let full_string = String::from("Hello, World!");
let slice = &full_string[0..5];
```

Here, `&full_string[0..5]` creates a reference to the first five characters of `full_string`.

> 
> A certain familiarity with the described methods and operators as well as the concept of string slices is crucial to
> work efficiently with and fully grasp the syntax of **Rust strings**.

Reviewer 2: ```suggestion
```
